File: BAEKJOON/그래프 탐색/아기 상어/아기 상어.py
from collections import deque

def bfs(start_x, start_y, size, grid):
    N = len(grid)
    # 네 방향(상, 하, 좌, 우)을 나타내는 리스트
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  
    # BFS를 위한 큐 초기화, 초기 위치와 거리를 큐에 추가
    queue = deque([(start_x, start_y, 0)])  
    # 방문한 위치를 저장하기 위한 집합
    visited = set()
    visited.add((start_x, start_y))
    # 먹을 수 있는 물고기 리스트 초기화
    fish = []

    while queue:
        x, y, dist = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < N and 0 <= ny < N and (nx, ny) not in visited:
                # 상어가 지나갈 수 있거나 먹을 수 있는 경우
                if grid[nx][ny] <= size:  
                    visited.add((nx, ny))
                    queue.append((nx, ny, dist + 1))
                    # 상어가 먹을 수 있는 물고기
                    if 0 < grid[nx][ny] < size:  
                        fish.append((dist + 1, nx, ny))
    
    # 먹을 수 있는 물고기가 없으면 None 반환
    if not fish:
        return None
    # 거리, 행, 열 순으로 정렬하여 가장 가까운 물고기 반환
    fish.sort()
    return fish[0]  # (거리, x, y)

def baby_shark(grid):
    N = len(grid)
    # 상어의 초기 위치를 찾고 grid에서 제거
    for i in range(N):
        for j in range(N):
            if grid[i][j] == 9:
                shark_x, shark_y = i, j
                grid[i][j] = 0

    size = 2
    eaten = 0
    time = 0

    while True:
        # BFS를 사용하여 먹을 수 있는 물고기를 찾음
        result = bfs(shark_x, shark_y, size, grid)
        if not result:
            break
        dist, shark_x, shark_y = result
        time += dist
        grid[shark_x][shark_y] = 0
        eaten += 1
        # 상어가 현재 크기만큼 물고기를 먹으면 크기 증가
        if eaten == size:
            size += 1
            eaten = 0

    return time

# 입력 받기
N = int(input().strip())
grid = [list(map(int, input().split())) for _ in range(N)]

# 결과 출력
print(baby_shark(grid))

File: BAEKJOON/그래프 탐색/아기 상어/아기 상어.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <algorithm>

using namespace std;

struct Fish {
    int dist;
    int x;
    int y;
};

bool compare(const Fish &a, const Fish &b) {
    if (a.dist == b.dist) {
        if (a.x == b.x) {
            return a.y < b.y;
        }
        return a.x < b.x;
    }
    return a.dist < b.dist;
}

vector<Fish> bfs(int start_x, int start_y, int size, vector<vector<int>> &grid) {
    int N = grid.size();
    vector<vector<bool>> visited(N, vector<bool>(N, false));
    queue<tuple<int, int, int>> q;
    q.push(make_tuple(start_x, start_y, 0));
    visited[start_x][start_y] = true;
    vector<Fish> fish;

    int directions[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    while (!q.empty()) {
        int x, y, dist;
        tie(x, y, dist) = q.front();
        q.pop();

        for (auto direction : directions) {
            int nx = x + direction[0];
            int ny = y + direction[1];

            if (nx >= 0 && nx < N && ny >= 0 && ny < N && !visited[nx][ny]) {
                if (grid[nx][ny] <= size) {
                    visited[nx][ny] = true;
                    q.push(make_tuple(nx, ny, dist + 1));
                    if (grid[nx][ny] > 0 && grid[nx][ny] < size) {
                        fish.push_back({dist + 1, nx, ny});
                    }
                }
            }
        }
    }

    sort(fish.begin(), fish.end(), compare);
    return fish;
}

int baby_shark(vector<vector<int>> &grid) {
    int N = grid.size();
    int shark_x, shark_y;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            if (grid[i][j] == 9) {
                shark_x = i;
                shark_y = j;
                grid[i][j] = 0;
            }
        }
    }

    int size = 2;
    int eaten = 0;
    int time = 0;

    while (true) {
        vector<Fish> result = bfs(shark_x, shark_y, size, grid);
        if (result.empty()) {
            break;
        }

        Fish nearest = result[0];
        time += nearest.dist;
        shark_x = nearest.x;
        shark_y = nearest.y;
        grid[shark_x][shark_y] = 0;
        eaten++;

        if (eaten == size) {
            size++;
            eaten = 0;
        }
    }

    return time;
}

int main() {
    int N;
    cin >> N;
    vector<vector<int>> grid(N, vector<int>(N));

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cin >> grid[i][j];
        }
    }

    cout << baby_shark(grid) << endl;
    return 0;
}
